src/main/java/com/iexec/core/replicate/ReplicateSupplyService.java

```diff
@@ -128,10 +128,17 @@ public ReplicateSupplyService(ReplicatesService replicatesService,
 
         for (Task task : validTasks) {
             String chainTaskId = task.getChainTaskId();
+            final Optional<ReplicatesList> oReplicatesList = replicatesService.getReplicatesList(chainTaskId);
+            // If there's no replicatesList, the task is probably not initialized.
```

Comment:
```suggestion
            // If there's no replicatesList, the task is probably not initialized.
            // FYI: initialized but not-yet-running tasks will have a defined ReplicateList with size == 0.
```

Reply:
This filtering should probably be done in the method `taskService.getInitialiazedOrRunningTasks()`.

Reply:
We can add a TODO comment to filter those tasks in-db directly if it is possible with mongo.

Reply:
I disagree. As @jbern0rd pointed out a few weeks ago, there could be some desynchronization there so we should check the current status of the replicates in the `for`-loop, the nearest to its use.

Reply:
The method `taskService.getInitialiazedOrRunningTasks()` returns non-initialized non-running tasks. Do you find that coherent?

Reply:
When they are retrieved, they are initialized or running. But their status can change between their retrieval and their use.
However, this filtering should never be used as `replicatesList` should not be back to an empty list. That's just a safeguard.

Reply:
Why not adding `ReplicatesList` retrieval in `runningTasks.stream()` filter for now?

Reply:
We need the freshest value of `ReplicatesList` in the `while` loop, so it has to be retrieved when the task is handled, not before.

Reply:
Yeah, I understand, but this method needs at least couple updates if we don't want to move everything in DB query for this PR:
1. `isTeeTask && !worker.isTeeEnabled()` can be done with DB query
2. `hasWorkerAlreadyParticipated()` should be static
3. `isInContributionPhase` is useless if code is fast
4. `smsService.getEnclaveChallenge(chainTaskId, isTeeTask)` call can be moved outisde plus sms service should cache enclave challenges to avoid repeating calls on same tasks

Reply:
I agree with you, there's still a lot to do on that method and the methods it calls. I'd say:
- every DB query that's needed to compute whether the worker can take the task or not should be done at the start of the `while` loop so that's fixed throughout the computation;
- consequence: every called method after the start of the `while` loop should be static.
Problem: using static methods is usually a bad practice. In particular, it makes things harder to unit test.

So, regarding your proposals:
1. Not sure what you mean. Do you mean the worker's state should be retrieved when needed - probably at the start of the `while` loop instead of before the loop? By the way, there's a TODO there: `access worker TEE status through workerService to avoid workerService.getWorker call`.
2. I don't think it'd be a good idea because of first paragraph. But it could be relieved from its call to the database.
3. Code is not always fast, remember @jbern0rd's work on fixing some `TASK_NOT_ACTIVE` issues. It usually depends on the load on the Scheduler. I don't think it'd be a good idea to remove that line.
4. Totally true. Let's use some memoization there!

To conclude: there's a lot to do to avoid race conditions there. We should probably also strive to propose some non-database-dependent methods in `ReplicateService` so that we would tempted to not use the database if not needed in future developments. See https://github.com/iExecBlockchainComputing/iexec-core/pull/490#discussion_r786713900 for an example.

Reply:
I see your point! Let's add this :)

Reply:
Added in https://github.com/iExecBlockchainComputing/iexec-core/pull/490/commits/50cb765c8117cb98b63fcbceddbbafc408e13d44.

Reply:
I suggest we swap from:
```
updateTaskStatusAndSave(task, INITIALIZED, null);
replicatesService.createEmptyReplicateList(task.getChainTaskId());
```
to:
```
replicatesService.createEmptyReplicateList(task.getChainTaskId());
updateTaskStatusAndSave(task, INITIALIZED, null);
```
and update current comment (`// If there's no replicatesList`) to express empty optional for ReplicateList cannot happen, check is only here to prevent `'Optional.get()' without 'isPresent()' check`. warning.

Reply:
Sure, good idea. See https://github.com/iExecBlockchainComputing/iexec-core/pull/490/commits/802aada1c3f0ed55ebf320170f8200d59f48c63c.